### PR TITLE
Implement Derived SubstateIds

### PR DESCRIPTION
**Pull request by talekhinezh**

### CODE REVIEW

File: simulator/src/ledger/radix_engine_db.rs [hunk 1]
```diff
@@ -94,51 +94,54 @@ impl QueryableSubstateStore for RadixEngineDB {
...
+impl WriteableSubstateStore for RadixEngineDB {
+    fn put_substate(&mut self, address: &[u8], substate: Substate) {
+        self.write(address, &scrypto_encode(&substate));
+    }
+
+    fn put_child_substate(&mut self, address: &[u8], key: &[u8], substate: Substate) {
+        let mut id = address.to_vec();
+        id.extend(key.to_vec());
+        self.write(&id, &scrypto_encode(&substate));
+    }
```
Comment (iamyulong) [inline simulator/src/ledger/radix_engine_db.rs:134]: Feel like we should get rid of the `put_child_substate` interface. It's causing very inconsistent physical keys in the database.

File: radix-engine/src/ledger/traits.rs [hunk 2]
```diff
@@ -124,6 +71,37 @@ pub trait SubstateStore {
...
+    fn get_nonce(&self) -> u64;
+}
+
+pub trait WriteableSubstateStore {
+    fn put_substate(&mut self, address: &[u8], substate: Substate);
+    fn put_child_substate(&mut self, address: &[u8], key: &[u8], substate: Substate);
+
+    fn put_keyed_substate(&mut self, address: &[u8], value: Vec<u8>, phys_id: (Hash, u32)) {
+        self.put_substate(address, Substate { value, phys_id });
+    }
```
Comment (iamyulong) [inline radix-engine/src/ledger/traits.rs:89]: Didn't see the difference between `put_keyed_substate` and `put_substate`. 

I'm supposing that there will be some semantic difference in the next PR:
-  One uses conflict free substate id
-  The other uses derived substate id where existence check is required (if so, the interface should return a result).

**Review by iamyulong — COMMENTED** [review]
The clean-up is nice

File: simulator/src/ledger/radix_engine_db.rs (hunk 1 above)
Comment (talekhinezh) [reply simulator/src/ledger/radix_engine_db.rs:134]: Yup agreed, refining this interface will be the subject of the next few PRs :)

File: radix-engine/src/ledger/traits.rs (hunk 2 above)
Comment (talekhinezh) [reply radix-engine/src/ledger/traits.rs:89]: Yup, this is mostly temporary interface in fact.

**Review by iamyulong — APPROVED** [review]
Let's go!

**Pull request merged**